File: csharp/ReplicatedFlag.cs
// A deliberately buggy workload, used to show deterministic-sim-testing finding +
// shrinking a bug.
//
// A coordinator sets a flag on two replicas with a single fire-and-forget message
// each - and, crucially, never retries. That missing retry is the bug: if the message
// to one replica is lost, the replicas diverge forever and nobody notices.

using System;
using System.Collections.Generic;
using System.Linq;

namespace SeedSim;

public sealed class Coordinator : Node
{
    private readonly List<string> _replicas;

    public Coordinator(List<string> replicas) : base("coord") => _replicas = replicas;

    public void Start()
    {
        // The bug in one line: send once, never confirm, never retry.
        foreach (var r in _replicas)
            Send(r, new Dictionary<string, object?> { ["kind"] = "set", ["value"] = 1 });
    }
}

public sealed class Replica : Node
{
    public int Value { get; private set; }

    public Replica(string nodeId) : base(nodeId) { }

    public override void OnMessage(string src, Dictionary<string, object?> msg)
    {
        if (Equals(msg["kind"], "set")) Value = (int)msg["value"]!;
    }
}

public static class ReplicatedFlag
{
    public static readonly string[] Nodes = { "coord", "r1", "r2" };
    public const int Horizon = 1000;

    public sealed record Result(List<TraceEntry> Trace, bool Diverged, Dictionary<string, int> Values);

    /// <summary>Run the workload once. Returns (trace, diverged, replica values).</summary>
    public static Result RunScenario(int seed, FaultSchedule faults)
    {
        var sim = new Simulator(seed, faults);
        var replicas = new List<Replica> { (Replica)sim.Add(new Replica("r1")), (Replica)sim.Add(new Replica("r2")) };
        var coord = (Coordinator)sim.Add(new Coordinator(replicas.Select(r => r.NodeId).ToList()));
        sim.At(0, coord.Start);
        sim.Run(until: Horizon);
        var values = replicas.ToDictionary(r => r.NodeId, r => r.Value);
        var diverged = values.Values.Distinct().Count() > 1;
        return new Result(sim.Trace, diverged, values);
    }

    /// <summary>Generate a small random fault schedule from a seed.</summary>
    public static List<Fault> RandomFaults(Rng rng, string[]? nodes = null)
    {
        nodes ??= Nodes;
        var outFaults = new List<Fault>();
        var count = rng.Between(1, 4);
        for (var i = 0; i < count; i++)
        {
            var start = (int)rng.Below(5);
            var end = start + (int)rng.Between(20, 200);
            if (rng.Chance(0.6))
            {
                var a = nodes[rng.Below(nodes.Length)];
                var b = nodes[rng.Below(nodes.Length)];
                if (a != b) outFaults.Add(new Partition(a, b, start, end));
            }
            else
            {
                var n = nodes[rng.Below(nodes.Length)];
                outFaults.Add(new Crash(n, start, end));
            }
        }
        return outFaults;
    }

    /// <summary>Return the first (seed, faults) whose run diverges, or null.</summary>
    public static (int Seed, List<Fault> Faults)? SearchForDivergence(IEnumerable<int> seeds)
    {
        foreach (var seed in seeds)
        {
            var faults = RandomFaults(new Rng(seed));
            var result = RunScenario(seed, FaultSchedule.Of(faults));
            if (result.Diverged) return (seed, faults);
        }
        return null;
    }
}

File: go/replicated_flag.go
// A deliberately buggy workload, used to show deterministic-sim-testing finding +
// shrinking a bug.
//
// A coordinator sets a flag on two replicas with a single fire-and-forget message
// each - and, crucially, never retries. That missing retry is the bug: if the
// message to one replica is lost, the replicas diverge forever and nobody notices.

package deterministicsimtesting

// Nodes are the participants in the worked example.
var Nodes = []string{"coord", "r1", "r2"}

// Horizon is the virtual time the scenario runs until.
const Horizon = 1000

// Coordinator sets a flag on each replica exactly once and never retries.
type Coordinator struct {
	BaseNode
	replicas []string
}

// NewCoordinator builds the coordinator for the given replica ids.
func NewCoordinator(replicas []string) *Coordinator {
	c := &Coordinator{replicas: replicas}
	c.ID = "coord"
	return c
}

// Start fires the coordinator's one-shot writes. The bug in one line: send once,
// never confirm, never retry.
func (c *Coordinator) Start() {
	for _, r := range c.replicas {
		c.Send(r, map[string]any{"kind": "set", "value": 1})
	}
}

// Replica stores a single flag value updated by "set" messages.
type Replica struct {
	BaseNode
	Value int
}

// NewReplica builds a replica with the given id and an initial value of 0.
func NewReplica(id string) *Replica {
	r := &Replica{}
	r.ID = id
	return r
}

// OnMessage applies a "set" message to the replica's value.
func (r *Replica) OnMessage(src string, msg map[string]any) {
	if msg["kind"] == "set" {
		r.Value = msg["value"].(int)
	}
}

// Result is the outcome of a single scenario run.
type Result struct {
	Trace    []TraceEntry
	Diverged bool
	Values   map[string]int
}

// Found is a seed together with the fault schedule that made it diverge.
type Found struct {
	Seed   int
	Faults []Fault
}

// RunScenario runs the workload once and reports the trace, whether the replicas
// diverged, and their final values.
func RunScenario(seed int, faults *FaultSchedule) Result {
	sim := NewSimulator(seed, faults)
	r1 := NewReplica("r1")
	sim.Add(r1)
	r2 := NewReplica("r2")
	sim.Add(r2)
	coord := NewCoordinator([]string{r1.NodeID(), r2.NodeID()})
	sim.Add(coord)
	sim.At(0, coord.Start)
	horizon := Horizon
	sim.Run(&horizon, 1_000_000)

	values := map[string]int{"r1": r1.Value, "r2": r2.Value}
	diverged := r1.Value != r2.Value
	return Result{Trace: sim.Trace, Diverged: diverged, Values: values}
}

// RandomFaults generates a small random fault schedule from a seeded Rng. Passing
// nil for nodes uses the default Nodes. Windows are biased to start early because
// all the interesting activity happens in the first few ticks.
func RandomFaults(rng *Rng, nodes []string) []Fault {
	if nodes == nil {
		nodes = Nodes
	}
	var out []Fault
	count := rng.Between(1, 4)
	for i := int64(0); i < count; i++ {
		start := int(rng.Below(5))
		end := start + int(rng.Between(20, 200))
		if rng.Chance(0.6) {
			a := nodes[rng.Below(int64(len(nodes)))]
			b := nodes[rng.Below(int64(len(nodes)))]
			if a != b {
				out = append(out, Partition{A: a, B: b, Start: start, End: end})
			}
		} else {
			n := nodes[rng.Below(int64(len(nodes)))]
			out = append(out, Crash{Node: n, Start: start, End: end})
		}
	}
	return out
}

// SearchForDivergence returns the first (seed, faults) whose run diverges, or nil.
func SearchForDivergence(seeds []int) *Found {
	for _, seed := range seeds {
		faults := RandomFaults(NewRng(int64(seed)), nil)
		if RunScenario(seed, FaultScheduleOf(faults)).Diverged {
			return &Found{Seed: seed, Faults: faults}
		}
	}
	return nil
}
